File: work/Smartlearning-Frontend/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  AdminCreateUserRequest,
  ChangePasswordRequest,
  UpdateUserRequest,
  UserResponse
} from '../models/models';

@Injectable({ providedIn: 'root' })
export class UserService {
  private readonly apiUrl = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) {}

  getMe(): Observable<UserResponse> {
    return this.http.get<UserResponse>(`${this.apiUrl}/me`);
  }

  getAllUsers(): Observable<UserResponse[]> {
    return this.http.get<UserResponse[]>(this.apiUrl);
  }

  getUser(id: string): Observable<UserResponse> {
    return this.http.get<UserResponse>(`${this.apiUrl}/${id}`);
  }

  getStudents(): Observable<UserResponse[]> {
    return this.http.get<UserResponse[]>(`${this.apiUrl}/students`);
  }

  getLecturers(): Observable<UserResponse[]> {
    return this.http.get<UserResponse[]>(`${this.apiUrl}/lecturers`);
  }

  createUserByAdmin(request: AdminCreateUserRequest): Observable<UserResponse> {
    return this.http.post<UserResponse>(`${this.apiUrl}/admin/create`, request);
  }

  /** Re-sends the invitation email for a still-pending account. */
  resendInvitation(id: string): Observable<string> {
    return this.http.post(`${this.apiUrl}/${id}/resend-invitation`, null, {
      responseType: 'text'
    });
  }

  updateUser(email: string, request: UpdateUserRequest): Observable<UserResponse> {
    return this.http.put<UserResponse>(`${this.apiUrl}/${email}`, request);
  }

  changePassword(email: string, request: ChangePasswordRequest): Observable<string> {
    return this.http.put(`${this.apiUrl}/change-password/${email}`, request, {
      responseType: 'text'
    });
  }

  deleteUser(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /** Downloads a CSV of users, optionally filtered to a single role. Pass null/undefined for everyone. */
  exportUsersCsv(role?: string | null): Observable<Blob> {
    const url = role ? `${this.apiUrl}/export?role=${role}` : `${this.apiUrl}/export`;
    return this.http.get(url, { responseType: 'blob' });
  }
}

File: work/SmartLearning-Backend/src/main/java/Final/year/project/SmartLearning/Users/UserService.java
package Final.year.project.SmartLearning.Users;

import Final.year.project.SmartLearning.dto.AdminCreateUserRequest;
import Final.year.project.SmartLearning.dto.ChangePasswordRequest;
import Final.year.project.SmartLearning.dto.UpdateUserRequest;
import Final.year.project.SmartLearning.dto.UserResponse;

import java.util.List;
import java.util.UUID;

public interface UserService {

    UserResponse getUser(UUID id);

    UserResponse getUserByEmail(String email);

    List<UserResponse> getAllUsers();

    /**
     * Deletes a user. ADMIN may delete STUDENT/LECTURER accounts; deleting an
     * ADMIN or MASTER account is reserved for MASTER. callerRole is the role
     * of whoever is making this request — passed in because Spring's
     * @PreAuthorize can only see the caller's own role, not the role of the
     * user being acted on.
     */
    void deleteUser(UUID id, Role callerRole);

    UserResponse createUserByAdmin(AdminCreateUserRequest request);

    /**
     * Rotates the invitation code of a still-pending account and re-sends
     * the invitation email. Fails if the account is already activated.
     */
    void resendInvitation(UUID id);

    void changePassword(String email, ChangePasswordRequest request);
    UserResponse updateUser(String email, UpdateUserRequest request);
    List<UserResponse> getUsersByRole(Role role);

    /** CSV export, optionally filtered to one role. Pass null for all users. */
    String exportUsersCsv(Role roleFilter);
}